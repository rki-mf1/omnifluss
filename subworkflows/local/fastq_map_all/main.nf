
include { FASTQ_ALIGN_BWA        } from '../../nf-core/fastq_align_bwa/main'
include { SAMTOOLS_FAIDX         } from '../../../modules/nf-core/samtools/faidx/main'
include { BAM_MARKDUPLICATES_PICARD } from '../../nf-core/bam_markduplicates_picard/main'

workflow FASTQ_MAP_ALL {
    take:
    tools           // string
    ch_reads        // channel: [ val(meta), fastq ]
    ch_ref          // channel: [ val(meta), fasta ]
    ch_bwa_index    // channel: [ val(meta), index ]

    main:
    ch_bam      = Channel.empty()
    ch_bai      = Channel.empty()
    ch_versions = Channel.empty()
    //for multiqc
    ch_multiqc_files = Channel.empty()

    if (tools.split(',').contains('bwa')) {

        FASTQ_ALIGN_BWA(
            ch_reads,       // channel: [ val(meta), fastq ]
            ch_bwa_index,   // channel: [ val(meta), index ]
            true,
            ch_ref          // channel: [ val(meta), fasta ]
        )
        ch_bam          = FASTQ_ALIGN_BWA.out.bam
        ch_bai          = FASTQ_ALIGN_BWA.out.bai
        ch_versions     = ch_versions.mix(FASTQ_ALIGN_BWA.out.versions.first())
    }

    if (tools.split(',').contains('picard_remove_duplicates')) {
        BAM_MARKDUPLICATES_PICARD(
            ch_bam,         // channel: [ val(meta), bam ]
            [[], []],
            [[], []]
        )
        ch_bam      = BAM_MARKDUPLICATES_PICARD.out.bam
        ch_bai      = BAM_MARKDUPLICATES_PICARD.out.bai
        ch_versions = ch_versions.mix(BAM_MARKDUPLICATES_PICARD.out.versions.first())
        //MultiQC
        ch_samtools_stats = BAM_MARKDUPLICATES_PICARD.out.stats.map{it[1]}
        ch_samtools_flagstat = BAM_MARKDUPLICATES_PICARD.out.flagstat.map{it[1]}
        ch_samtools_idxstats = BAM_MARKDUPLICATES_PICARD.out.idxstats.map{it[1]}
        ch_picard_markduplicate_metrics = BAM_MARKDUPLICATES_PICARD.out.metrics.map{it[1]}
        //combining channels
        ch_multiqc_files = ch_samtools_stats.mix(ch_samtools_flagstat,ch_samtools_idxstats,ch_picard_markduplicate_metrics).view()
    }

    emit:
    bam           = ch_bam
    bai           = ch_bai
    versions      = ch_versions
    //MultiQC
    multiqc_files       = ch_multiqc_files


}

