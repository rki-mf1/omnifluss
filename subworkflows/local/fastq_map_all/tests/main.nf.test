nextflow_workflow {

    name "Test Subworkflow FASTQ_MAP_ALL"
    script "../main.nf"
    // Incude a nextflow.config for the test if needed
    config "subworkflows/local/fastq_map_all/tests/nextflow.config"
    workflow "FASTQ_MAP_ALL"

    //  Add tags for the subworkflow
    tag "subworkflows"
    tag "subworkflows/local"
    tag "subworkflows/fastq_map_all"
    tag "fastq_map_all"
    tag "subworkflows_nfcore"
    tag "subworkflows/fastq_align_bwa"
    tag "fastq_align_bwa"
    tag "subworkflows/bam_sort_stats_samtools"
    tag "bam_sort_stats_samtools"
    tag "subworkflows/bam_markduplicates_picard"
    tag "bam_markduplicates_picard"
    tag "subworkflows/bam_stats_samtools"
    tag "bam_stats_samtools"
    // Add tags for all modules used within this subworkflow.
    tag "bwa_mem"
    tag "samtools_sort"
    tag "samtools_index"
    tag "samtools_flagstat"
    tag "samtools_idxstats"
    tag "samtools_stats"
    tag "samtools/faidx"
    tag "picard/markduplicates"

    // Test name preferably indicating the test-data and file-format used
    test("sarscov2 - single_end - stub") {

        // Activate the stub mode for nextflow
        options '-stub'

        when {
            workflow {
                """
                input[0] = 'bwa,picard_remove_duplicates'
                input[1] = Channel.of([
                    [ id:'test', single_end:true ], // meta map
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true) ]
                    ])
                input[2] = Channel.value([
                    [id:'test'], //meta map
                    file("$projectDir/assets/dummy_file.txt", checkIfExists: true)
                ])
                input[3] = Channel.of([
                    [id:'test'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                    ])
                input[4] = Channel.of([
                    [id:'test'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)
                    ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("sarscov2 - paired_end - stub") {

        // Activate the stub mode for nextflow
        options '-stub'

        when {
            workflow {
                """
                input[0] = 'bwa,picard_remove_duplicates'
                input[1] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    [ file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true) ]
                    ])
                input[2] = Channel.value([
                    [id:'test'], //meta map
                    file("$projectDir/assets/dummy_file.txt", checkIfExists: true)
                ])
                input[3] = Channel.of([
                    [id:'test'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                    ])
                input[4] = Channel.of([
                    [id:'test'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)
                    ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
