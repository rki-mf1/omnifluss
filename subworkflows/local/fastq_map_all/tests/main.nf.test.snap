{
    "sarscov2 - single_end - stub": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "bai": [
                    
                ],
                "bam": [
                    
                ],
                "multiqc_files": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.1"
        },
        "timestamp": "2025-04-11T10:39:10.044617628"
    },
    "sarscov2 - paired_end - stub": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "bai": [
                    
                ],
                "bam": [
                    
                ],
                "multiqc_files": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.1"
        },
        "timestamp": "2025-04-11T10:39:16.98150977"
    }
}